File: day_02/part_one.py
limits = {"red": 12, "green": 13, "blue": 14}


def parse_line(line):
    chunk = line.strip().split(":")
    id = int(chunk[0].split(" ")[-1])
    cube_sets = chunk[1].split(";")

    for set in cube_sets:
        color_sets = set.split(",")
        for color_set in color_sets:
            num_color_cube = color_set.strip().split(" ")
            num = int(num_color_cube[0])
            color = num_color_cube[1]
            if num > limits[color]:
                id = 0

    # Woops! Answers a different question...
    # If the cumulative number (summation) of cubes used in each
    # Game (not round) falls within the limit

    # rgb = {"red": 0, "green": 0, "blue": 0}
    # for set in cube_sets:
    #     color_sets = set.split(",")
    #     for color_set in color_sets:
    #         num_color_cube = color_set.strip().split(" ")
    #         num = int(num_color_cube[0])
    #         color = num_color_cube[1]
    #         rgb[color] += num

    # for key in rgb.keys():
    #     if rgb[key] > limits[key]:
    #         id = 0

    return id


def main():
    with open("input") as input:
        sum = 0
        for line in input:
            sum += parse_line(line)
        print(f"The sum is: {sum}")


if __name__ == "__main__":
    main()

# The sum is: 2505

File: day_02/PartOne.java
package org.adventofcode2023;

import lombok.SneakyThrows;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;

public class PartOne {

    public static Map<String, Integer> limits = new HashMap<>() {{
        put("red", 12);
        put("green", 13);
        put("blue", 14);
    }};

    public static int parseLine(String line) {
        String[] chunk = line.split(":");
        int id = Integer.parseInt(chunk[0].split(" ")[chunk[0].split(" ").length - 1]);
        String[] cubeSets = chunk[1].split(";");

        for (String set : cubeSets) {
            String[] colorSets = set.split(",");
            for (String colorSet : colorSets) {
                String[] numColorCube = colorSet.trim().split(" ");
                int num = Integer.parseInt(numColorCube[0]);
                String color = numColorCube[1];
                if (num > limits.get(color)) {
                    id = 0;
                }
            }
        }

        return id;

    }

    public static void main(String[] args) {
        var input = getInput("input");

        int sum = 0;
        String[] inputLines = input.split("\n");
        for (String line : inputLines) {
            sum += parseLine(line);
        }

        System.out.printf("The sum is: %s", sum);
    }

    @SneakyThrows
    @SuppressWarnings({"SameParameterValue", "DataFlowIssue"})
    private static String getInput(String fileName) {
        return Files.readString(Paths.get(Main.class.getClassLoader().getResource(fileName).toURI())).trim();
    }
}